src/Syntax/GrammarRepository.re

```diff
@@ -25,18 +25,33 @@ let getGrammar = (~scope: string, gr: t) => {
     switch (Ext.LanguageInfo.getGrammarPathFromScope(gr.languageInfo, scope)) {
     | Some(grammarPath) =>
       gr.log("Loading grammar from: " ++ grammarPath);
-      let json = Yojson.Safe.from_file(grammarPath);
-      let resultGrammar = Textmate.Grammar.Json.of_yojson(json);
-
-      switch (resultGrammar) {
-      | Ok(grammar) =>
-        gr.log("Grammar loaded successfully");
-        Hashtbl.add(gr.scopeToGrammar, scope, grammar);
-        Some(grammar);
-      | Error(e) =>
-        gr.log("Grammar loading failed with: " ++ e);
-        None;
+
+      switch (Yojson.Safe.from_file(grammarPath)) {
+      | json =>
+        switch (Textmate.Grammar.Json.of_yojson(json)) {
+        | Ok(grammar) =>
+          gr.log("JSON Grammar loaded successfully");
+          Hashtbl.add(gr.scopeToGrammar, scope, grammar);
+          Some(grammar);
+
+        | Error(e) =>
+          gr.log("Grammar loading failed with: " ++ e);
+          None;
+        }
+
+      | exception (Yojson.Json_error(_)) =>
```

Comment:
I wonder if we should determine this based on file extension (or - at the very least - decide which implementation to run first?)

Reply:
I've just assumed they could have any extension, although we could of course still try to infer it from some common extensions. But does it really matter? Do you think there's really a noticeable effect from just trying it blindly?

Reply:
> But does it really matter? Do you think there's really a noticeable effect from just trying it blindly?

The main concern I had was performance related to synchronous file I/O - with the current strategy, for XML grammars, we always have to load them twice from the file system, and some of the grammars can be fairly large (~200Kb or so). 

That's my only concern; if there's no perf impact with this approach, I'm fine with it.

Reply:
They're cached though, so it should at least only happen once per grammar. And the parsing should fail pretty quickly, so unless the JSON parser reads the entire file in before parsing, file size shouldn't matter. The XML parser definitely doesn't, as it's a streaming API, but I also think it's so fault-tolerant that it'll probably just return the entire JSON file as a text node if we switch it around.

I can check what Yojson does though.

Reply:
Yojson uses ocamllex, which I'm pretty sure reads incrementally.

Reply:
Thanks for looking into this. If they are both streaming / incremental, should be negligible overhead.

Reply:
The overhead would also only be for XML grammars, which most aren't.